core/src/main/java/kafka/server/share/SharePartition.java

```diff
@@ -1058,36 +1084,57 @@ private void maybeCompleteInitialization(CompletableFuture<Void> future) {
 
     private AcquiredRecords acquireNewBatchRecords(
         String memberId,
+        Iterable<? extends RecordBatch> batches,
         long firstOffset,
-        long lastOffset
+        long lastOffset,
+        int maxFetchRecords
     ) {
         lock.writeLock().lock();
         try {
+            // If same batch is fetched and previous batch is removed from the cache then we need to
+            // update the batch first offset to endOffset, only if enfOffset is passed the firstOffset.
+            // For an initial start of the share fetch from a topic partition the endOffset will be initialized
+            // to 0 but firstOffset can be higher than 0.
+            long firstAcquiredOffset = firstOffset;
+            if (cachedState.isEmpty() && endOffset > firstAcquiredOffset) {
+                firstAcquiredOffset = endOffset;
```

Comment:
The situation where we need to move the firstAcquiredOffset is only when we receive first ever shar partition acquire request. In that case, cache shall be empty. Also we should only move the offset if endOffset is passed the firstOffset.

Reply:
Is that true? Suppose endOffset is initialized to 105 and the log has two batches with offset range [100, 110] and [111, 120]. Initially cachedState is empty and we call acquireNewBatchRecords with the first batch. We will set firstAcquiredOffset to 105 in this case. [105, 110] will be in acquired state and endOffset advances to 111. Another share fetch request comes in and we call acquireNewBatchRecords with the second batch. [111, 120] will be in acquired state and endOffset advances to 121. The first batch times out. [105, 110] will be in available state and endOffset is reset to 105. Now another share fetch request comes in. We will call acquireNewBatchRecords with the first batch. Since cachedState is not empty, we won't set firstAcquiredOffset to 105 and leave it as 100, which seems incorrect.

Reply:
> Since cachedState is not empty, we won't set firstAcquiredOffset to 105 and leave it as 100, which seems incorrect.

Yes, that's fine. As in those cases we need not to move firstAcquired offset as we are not handling fresh batch acquisition i.e. this check in only important when we are acquiring the batch for first time.

This method will not be triggered as submap is not empty, i.e. replica manager batch [100, 100] will have an overlap with [105, 110] cached batch.

The complete batch from [105,110] will be acquired as per the lines from 633, where inflight batch is a full match and available to acquire.